starboard/nplb/maximum_player_configuration_explorer.h

```diff
@@ -0,0 +1,141 @@
+
+// Given a vector v, test if it is contained within a max elements set s.
+// v is contained in a max elements set s if and only if there is a max element
+// m within s such that m >= v.
+template <typename VectorContainer>
+bool PosetContainedByMaxElementsSet(const VectorContainer& max_elements_set,
+                                    const std::vector<int>& v) {
+  for (const auto& max_element : max_elements_set) {
+    if (PosetGreaterThanOrEqualTo(max_element, v)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Given a vector, test if it is excluded from a max elements set.
+template <typename VectorContainer>
+bool PosetNotContainedByMaxElementsSet(const VectorContainer& max_elements_set,
+                                       const std::vector<int>& v) {
+  return !PosetContainedByMaxElementsSet<VectorContainer>(max_elements_set, v);
+}
+
+// The scheduler is responsible for arranging the search process to identify the
+// maximum elements within a graded poset, which is based on a set of vectors
+// comprising integers. Specifically, the length of each vector is determined by
+// the number of |resource_types|, while the range of integers in each vector
+// spans from 0 to the maximum number of |max_instances|, as [0, max_instances].
+// The partial order <= between two vectors is defined such that one vector is
+// considered less than or equal to the other vector if all the corresponding
+// integers in the first vector are smaller than or equal to the corresponding
+// integers in the second vector.
+// To identify the maximum elements, a test function, denoted as |test_functor|,
+// must be provided to evaluate whether the current configuration satisfies the
+// criterion for a maximum element.
+// A termination function |terminate_functor|, if provided, serves to terminate
+// the search process once the specified stop criterion is met. It is commonly
+// used to limit the search space by imposing an upper bound on the total number
+// of created instances.
+// The reason for applying the Depth First Search scheduler in this context is
+// that it is a faster method.
+std::set<std::vector<int>> ScheduleDFSOnGradedPoset(
+    int resource_types,
+    int max_instances,
+    const std::function<bool(const std::vector<int>&)>& test_functor,
+    const std::function<bool(const std::vector<int>&)>& terminate_functor =
+        nullptr);
+
+// The |MaximumPlayerConfigurationExplorer| is employed to evaluate the maximum
+// feasible configurations for a group of players, with each player using a
+// distinct video/audio codec and output mode.
+// The search process is carried out using the |ScheduleDFSOnGradedPoset|, while
+// the |Create| function is provided as the test function.
+class MaximumPlayerConfigurationExplorer {
+ public:
+  typedef std::tuple<
+      SbMediaVideoCodec,
+      std::pair<SbMediaAudioCodec,
+                const shared::starboard::media::AudioStreamInfo*>,
```

Comment:
Consider passing in `AudioStreamInfo` instead of `AudioStreamInfo*` as `AudioStreamInfo` is copyable.

Reply:
We can change it but please be aware of 2 things.

1. It is necessary to provide a comparator for this tuple, specifically for AudioStreamInfo. Currently, we don't have one in place because the enum and pointer types have default comparators.

2. If that's the case, we should consider returning AudioStreamInfo directly for the following API.
```
  const media::AudioStreamInfo& audio_stream_info() const {
    return dmp_info_.audio_sample_info.stream_info;
  }
```

Finally, it seems that implementing step 1 does not simplify the flow anymore.